### DISCUSSION TITLE
CUDA error: device-side assert triggered and modelling a Dependency Parse

### DISCUSSION THREAD

**Discussion by afonso-sousa** [pyg-team/pytorch_geometric#5663] [Q&A] [answered]
I am trying to use GNNs to learn a representation of a dependency parse. I have the following data batch: `DataBatch(x=[652, 1], edge_index=[2, 652], batch=[652], ptr=[33])`, where each node has a single feature (the corresponding word tokenized). The edge indices correspond to a Levy Graph (the relation types were transformed to nodes so the graph has no edge types, for simplicity). I have a simple network borrowed from a tutorial with a global pooling layer at the end to have a representation per batch.
```
class GCNEncoder(nn.Module):
    def __init__(self, num_features, hidden_size, dropout=0):
        super().__init__()
        self.conv1 = GCNConv(num_features, hidden_size)
        self.conv2 = GCNConv(hidden_size, hidden_size)
        self.relu = nn.ReLU(inplace=True)
        self.dropout = nn.Dropout(p=dropout)

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = self.conv1(x, edge_index)
        x = self.relu(x)
        x = self.dropout(x)
        x = self.conv2(x, edge_index)
        x = self.relu(x)
        x = global_add_pool(x, data.batch, size=data.num_graphs)
        return x
```
I am getting `RuntimeError: CUDA error: device-side assert triggered`. I do not compute a loss so what do you think may be causing the problem? Thanks in advance for any help you can provide.

PS: Additionally, this problem would be better suited for a RGAT, but I did not figure out what was "edge type" supposed to be.

**Comment by EdisonLeeeee** [comment]
I think the problem might be `x = global_add_pool(x, data.batch, size=data.num_graphs)`. Can you try this instead?
```python
x = global_add_pool(x, data.batch)
```

**Reply by afonso-sousa** [reply]
It did not work. Thanks for the reply anyways.

**Comment by rusty1s** [comment, marked as answer]
Can you run your code via `CUDA_LAUNCH_BLOCKING=1 python ...`? In which module does your script crash?

**Reply by afonso-sousa** [reply]
The stack trace shows the following:
```
...
graph_embeddings = self.graph_encoder(graph)
  File "/home/admin/anaconda3/envs/catbird/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/admin/projects/catbird/catbird/models/encoders/gcn_encoder.py", line 20, in forward
    x = self.conv1(x, edge_index)
  File "/home/admin/anaconda3/envs/catbird/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1110, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/admin/anaconda3/envs/catbird/lib/python3.8/site-packages/torch_geometric/nn/conv/gcn_conv.py", line 194, in forward
    out = self.propagate(edge_index, x=x, edge_weight=edge_weight,
  File "/home/admin/anaconda3/envs/catbird/lib/python3.8/site-packages/torch_geometric/nn/conv/message_passing.py", line 309, in propagate
    coll_dict = self.__collect__(self.__user_args__, edge_index,
  File "/home/admin/anaconda3/envs/catbird/lib/python3.8/site-packages/torch_geometric/nn/conv/message_passing.py", line 202, in __collect__
    data = self.__lift__(data, edge_index, dim)
  File "/home/admin/anaconda3/envs/catbird/lib/python3.8/site-packages/torch_geometric/nn/conv/message_passing.py", line 172, in __lift__
    return src.index_select(self.node_dim, index)
RuntimeError: CUDA error: device-side assert triggered
```
Thanks for the reply.

**Reply by EdisonLeeeee** [reply]
Can you have a try on CPU? This should provide more detailed information about the error.

**Reply by rusty1s** [reply]
This seems related to an out-of-bound index error. Can you ensure that `edge_index.max() < data.x.size(0)`?

**Reply by afonso-sousa** [reply]
edge_index.max() is equal to 646 and data.x.size(0) is equal to 652. The only difference running in CPU is that it throws "IndexError: index out of range in self".

**Reply by afonso-sousa** [reply]
Can the issue have something to do with the actual values?
In every batch sample, my graph is a dependency parse in which I tokenize the words, so my x has integer values, like: tensor([[2.2640e+03],
        [2.7091e+04],
        [7.4240e+03],
        [1.0000e+02],
        [1.6111e+04],
        [5.6000e+02],
        [3.7362e+04],
        [4.7830e+03],...

My edge_index always starts with 0, like:
 tensor([[  1,  11,   4,  ..., 644, 627, 645],
        [ 11,   0,  12,  ..., 628, 645, 629]])

Do you see anything wrong with that?

**Reply by rusty1s** [reply]
This is weird. Can you also try to run `data.validate()` and see if it reports any errors? Other than that, you should consider normalizing your input features. These large values are not really suitable for ML tasks:
```python
x = (x - x.mean(dim=0, keepdim=True)) / x.std(dim=0, keepdim=True)

**Reply by afonso-sousa** [reply]
Trying to print `data.validate()` throws `AttributeError: 'GlobalStorage' object has no attribute 'validate'`. My batching is done with PyGDataLoader() and inside my torch.utils.data.Dataset I have a graph per sample created using `torch_geometric.data.Data`.
My batch prints as follows:
`DataBatch(x=[652, 1], edge_index=[2, 652], batch=[652], ptr=[33])
`

**Reply by EdisonLeeeee** [reply]
It looks like you are using an old version of PyG. Can you upgrade your PyG to the latest version and try it again? Perhaps the bug has been fixed :) Otherwise you can run `data.validate()` again to see if it reports any errors.

**Reply by afonso-sousa** [reply]
Upgrading the version did the trick. It showed a more specific error message and I figured that somehow I had an edge index of -1. Thank you very much for the help.
